planning/MartenV4.md

```diff
@@ -0,0 +1,405 @@
+
+The event sourcing in Marten has limited scalability because of its dependence today upon appending events to 
+one table. Closely related to the storage is ways to segment the asynchronous projection support to parallelize
+the work across multiple threads or active nodes.
+
+Here are some ideas:
+
+* Allow the user to define multiple event logs in the database. This way you'd have multiple stream/event tables for completely
+  different kinds of streams
+* Postgresql native partitioning by tenant id where that's useful
+* Aggregation keys -- this is tightly coupled to the projections. If there's a projection that is done by something other than tenant id or stream id, try to persist the 
+  key for however the events/streams are grouped to stream and/or event storage. Thinking about data elements like "Region" or "Country" where you might be doing
+  aggregations across streams or tenants. Maybe that's useful?
+* Archive streams that reach some kind of termination event? Move everything to an archive table off to the side? Keeps the event log table cleaner
+* Artificial segmenting of the streams. Use some kind of algorithm that could see stream ids, and assign them to 2 or more "cohorts", then persist that "cohort key" to 
+  both the stream and event data. This would potentially help for the async daemon to parallelize work. It would definitely help out for "live" projections by having 
+  smaller tables to work through
+
+### Projections
+
+Thoughts:
+
+Functionality wise, the lifecycle of projections are:
+
+* "Live" projections built on demand from the raw events. This only changes for v4 to take advantage of new snapshotting capabilities
+* "Inline" projections built as part of the same unit of work as capturing the source events. See the section on *Inline Projections* because this
+  will change internally for v4
+* "Async" projections built up in a background process of some sort. I think this turns into a near re-write for v4
+
+By *type*, the projection types *could* be:
+
+* Aggregate document by stream (existing)
+* Aggregate document by tenant id (new)
+* Execute SQL statement by event type (new) -- so, insert/update/delete. This comes up a lot, and it should be low hanging fruit
+* Transform a single event to a document --> needs to become just "carry out some kind of change to a document" and probably ends up sharing more mechanics with the aggregation than it used to
+* Aggregate by arbitrary values of either the stream or event types -- but if we do this, I strongly vote to persist whatever the differentiation value is in the database to make projection rebuilds and the async daemon more "parallelizable"
+
+
+Other thoughts:
+
+* I want to completely throw away the current `Aggregator` implementations. It was just too naive for real world usage and the internals have gotten ugly
+* Re-evaluate the `ViewProjection` fluent interface. That might fit in fine to our new model of projections for v4 with more overloads
```

Comment:
For me `ViewProjections` even though the API is not ideal was the 90% (or even more) usage of the projections.

Reply:
`ViewProjection` has turned into my "default" projection type as well. The place where it usually breaks down for me is that I occasionally need to project one event to several documents, which it ofc doesnt work for and then I usually have to drop down to the raw `IProjection` interface. 

Im not saying that `ViewProjection` needs to support that use case though. I think its just fine to have a thing that does one thing well.

Reply:
Fully agreed.

Btw. @wastaz how do you find usage of the `IProjection` interface?

Reply:
`IProjection` is cumbersome to work with tbh. Its nice to have the option, and its gotten a lot better since C# introduced pattern matching. But its not something I _want_ to work with in many cases. Also I still dont really understand how `Apply` vs `ApplyAsync` work. Id have expected them to be like the standard .net XYZ/XYZAsync method pairs but it _seems_ like they are actually part of the inline/async-projection split instead and...its confusing.

Reply:
Yes, that’s why I’m asking. I didn't want to suggest an answer, but IMHO we should provide something more useful (or simpler) for the users. 
This interface is similar to the Event Store subscription, so it should still be an option, but would be nice to provide an easier option (eg. with Event instead of the whole Event Page).

Reply:
@wastaz @oskardudycz I'd like to see if we could unify the projection definition through some mix of an improved `ViewProjection` and my *Transformer* idea and remove the `IProjection` interface altogether. 

And yeah, there's some mixed concerns with `IProjection` between things that only matter for inline projections and some things that only matter for async.